runtime/ts/storage/pouchdb/pouch-db-storage.ts

```diff
@@ -0,0 +1,245 @@
+// @
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {assert} from '../../../../platform/assert-web.js';
+import {StorageBase, StorageProviderBase} from '../storage-provider-base.js';
+import {PouchDbKey} from './pouch-db-key.js';
+import {Id} from '../../id.js';
+import {Type} from '../../type.js';
+import {PouchDbCollection} from "./pouch-db-collection.js";
+import {PouchDbStorageProvider} from "./pouch-db-storage-provider.js";
+import {PouchDbBigCollection} from "./pouch-db-big-collection.js";
+import {PouchDbVariable} from "./pouch-db-variable.js";
+
+import PouchDB from 'pouchdb';
+import PouchDbMemory from 'pouchdb-adapter-memory';
+
+
+export class PouchDbStorage extends StorageBase {
+  // TODO(lindner) add global weak map of keys and handle replication events.
+  private readonly remoteStateChangedHandlers: Map<string, PouchDbStorageProvider> = new Map();
+  
+  // Used for reference mode
+  private readonly baseStores: Map<Type, PouchDbCollection> = new Map();
+  private readonly baseStorePromises: Map<Type, Promise<PouchDbCollection>> = new Map();
+  private localIDBase: number;
+
+  /** Global map of database types/name to Pouch Database Instances */
+  private static dbLocationToInstance: Map<string, PouchDB.Database> = new Map();
+
+  /** Tracks replication status and allows cancellation in tests */
+  private static syncHandler: PouchDB.Replication.Sync<{}>|undefined;
+
+  constructor(arcId: Id) {
+    super(arcId);
+    this.localIDBase = 0;
+  }
+
+  /**
+   * Instantiates a new key for id/type stored at keyFragment.
+   */
+  async construct(id: string, type: Type, keyFragment: string) : Promise<PouchDbStorageProvider> {
+    const provider = await this._construct(id, type, keyFragment);
+    if (type.isReference) {
+      return provider;
+    }
+    if (type.isTypeContainer() && type.getContainedType().isReference) {
+      return provider;
+    }
+    provider.enableReferenceMode();
+    return provider;
+  }
+
+  async _construct(id: string, type: Type, keyFragment: string) {
+    const key = new PouchDbKey(keyFragment);
+    const keystr = key.toString();
+    
+    const provider = this.newProvider(type, undefined, id, key.toString());
+
+    // Used to track changes for the key.
+    this.remoteStateChangedHandlers.set(key.location, provider);
+    
+    return provider;
+  }
+
+  /**
+   * Connect with an existing storage key.  Returns a cached instance if available.
+   * Returns null if no such storage key exists.
+   */
+  async connect(id: string, type: Type, key: string) : Promise<PouchDbStorageProvider> {
+    const imKey = new PouchDbKey(key);
+
+    // TODO(lindner): fail if not created.
+    return this.construct(id, type, key);
```

Comment:
The semantic of construct is that it should fail if the storage key provided already points to data; and the semantic of connect is that it should fail if this is *not* true. So one shouldn't be calling the other.

Reply:
right, I realize this, and will add the functionality, it wasn't blocking my manual testing (yet)